pycqed/measurement/sweep_points.py

```diff
@@ -136,48 +159,144 @@ def get_sweep_params_property(self, property, dimension, param_names=None):
+        :return: dict of the form
+         {mobj_name: [sweep_param_name_0, ..., sweep_param_name_n]}
         """
-
-        if len(measured_objects) != len(self[0]):
-            raise ValueError('The number of keys and number of sweep '
-                             'parameters do not match.')
+        if not isinstance(measurement_objects, list):
+            measurement_objects = [measurement_objects]
+        for i, mobj in enumerate(measurement_objects):
+            if hasattr(mobj, 'name'):
+                measurement_objects[i] = mobj.name
 
         sweep_points_map = OrderedDict()
-        for i, key in enumerate(measured_objects):
-            sweep_points_map[key] = [list(d)[i] for d in self]
-
+        for i, mobjn in enumerate(measurement_objects):
+            sweep_points_map[mobjn] = []
+            for dim, d in enumerate(self):
+                if len(d) == 1:
+                    # assume all mobjs use the same param_name
+                    sweep_points_map[mobjn] += [next(iter(d))]
+                elif mobjn in list(d)[i]:
+                    sweep_points_map[mobjn] += [list(d)[i]]
+                else:
+                    if len(d) != len(measurement_objects):
+                        raise ValueError(
+                            f'{len(measurement_objects)} measurement objects '
+                            f'were given but there are {len(d)} '
+                            f'sweep parameters in dimension {dim}.')
+                    sweep_points_map[mobjn] += [list(d)[i]]
         return sweep_points_map
 
+    def length(self, dimension='all'):
+        """
+        Returns the number of sweep points in a given sweep dimension (after a
+        sanity checking).
+
+        :param dimension: ('all' or int) sweep dimension (default: 'all').
+
+        :return: (int) number of sweep points in the given dimension
+        """
+
+        if dimension == 'all':
+            return [self.length(d) for d in range(len(self))]
+
+        if len(self) == 0 or (dimension >= 0 and dimension >= len(self)):
+            return 0
+        n = 0
+        for p in self[dimension].values():
+            if n == 0:
+                n = len(p[0])
+            elif n != len(p[0]):
+                raise ValueError('The lengths of the sweep points are not '
+                                 'consistent.')
+        return n
+
+    def update(self, sweep_points):
+        """
+        Updates the sweep dictionaries of all dimensions with the sweep
+        dictionaries passed as sweep_points. Non-exisiting
+        parameters and required additional dimensions are added if needed.
+
+        :param sweep_points: (SweepPoints) a SweepPoints object containing
+            the sweep points to be updated.
+
+        :return:
+        """
+        while len(self) < len(sweep_points):
+            self.add_sweep_dimension()
+        for d, u in zip(self, sweep_points):
+            d.update(u)
+
+    def find_parameter(self, param_name):
```

Comment:
I think I would combine this function with get_sweep_dimension by adding an extra optional argument to the latter. I also find find_parameter a bit of a misnomer because it rather finds a dimension not a parameter.
What do you think?